### PR TITLE
Echo Server

### PR DESCRIPTION
**Pull request by xray**
Netcat Echo Server, all tests passing.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,11 +1,15 @@
-# Chorizo (C#/.NET) - v0.1  
+# Chorizo (C#/.NET) - v0.5
```
Comment (ssunday) [inline README.md]: This is the version of the software? Is there a changelog ?

File: src/ChorizoServer.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        public Chorizo()
+        {
+            ProtocolHandler = new CzoTelNetHandler();
+            SocketMachine = SocketMachine ?? new DotNetSocketMachine();
+            Config = new ServerConfig("HTTP", "localhost", 8000);
+        }
+
+        public Chorizo(int port, string protocol = null)
+        {
+            protocol = protocol ?? "HTTP";
```
Comment (ssunday) [inline src/ChorizoServer.cs]: There seems to be some duping between these blocks—can they be simplified?

File: src/SocketMachine/DotNetSocketMachineOld.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+//            var localEndPoint = new IPEndPoint(ipAddress, port);
+//
+//            SocketImplementation = SocketImplementation ?? new DotNetSocket();
+//            
+//            SocketImplementation.Bind(localEndPoint);
+//            SocketImplementation.Listen(backlog);
+//        }
+//
+//        public ICzoSocket AcceptConnection()
+//        {
```
[the remark this answers is not in the document]
Comment (xray) [reply src/SocketMachine/DotNetSocketMachineOld.cs]: Completely agree!

File: README.md (hunk 1 above)
Comment (xray) [reply README.md]: That is the intention, there is not a changelog at this time.

File: src/ChorizoServer.cs (hunk 2 above)
Comment (xray) [reply src/ChorizoServer.cs]: Should be better now!

File: src/ChorizoServer.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        public IServerStatus Status { get; set; }
+        public ISocketMachine SocketMachine;
+        public ICzoProtocolHandler ProtocolHandler;
+
+        private ServerConfig Config { get; }
+
+        public Chorizo(int port = 8000, string protocol = "HTTP")
+        {
+            ProtocolHandler = new CzoTelNetHandler();
```
Comment (cmvandrevala) [inline src/ChorizoServer.cs]: I would rename "Czo" to "Chorizo"

File: src/ChorizoServer.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+            SocketMachine = SocketMachine ?? new DotNetSocketMachine();
```
Comment (cmvandrevala) [inline src/ChorizoServer.cs]: Why is this called a `Machine`?

File: src/ChorizoServer.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            Config = new ServerConfig(protocol, "localhost", port);
+        }
+
+        public void Start()
+        {
+            SocketMachine.Setup(Config.Port, Config.HostName);
+            SocketMachine.Listen(Config.Backlog);
```
Comment (cmvandrevala) [inline src/ChorizoServer.cs]: You could conceivably call this "NumberOfConnectionsInQueue"

File: src/ChorizoServer.cs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+                    ProtocolHandler.Handle(chorizoSocket);
+                }
+                else
+                {
+                    throw new NotImplementedException();
+                }
+            }
+        }
+    }
+}
```
Comment (cmvandrevala) [inline src/ChorizoServer.cs]: There are plugins that you can install to automatically add a newline at the end of each file and clear excess whitespace

File: src/ProtocolHandler/CzoTelNetHandler.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
+using System;
+using System.Net.Sockets;
+using System.Text;
+using Chorizo.Sockets.CzoSocket;
+
+namespace Chorizo.ProtocolHandler
+{
+    public class CzoTelNetHandler : ICzoProtocolHandler
```
Comment (cmvandrevala) [inline src/ProtocolHandler/CzoTelNetHandler.cs]: Rename "Czo" to "Chorizo"

File: src/ProtocolHandler/CzoTelNetHandler.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        private readonly string _protocol;
```
Comment (cmvandrevala) [inline src/ProtocolHandler/CzoTelNetHandler.cs]: Declare the value of _protocol here

File: src/ProtocolHandler/CzoTelNetHandler.cs [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        public CzoTelNetHandler()
+        {
+            _protocol = "TelNet";
+        }
+        public bool WillHandle(string protocol)
+        {
+            return protocol == _protocol;
+        }
+
+        public void Handle(ICzoSocket chorizoSocket)
```
Comment (cmvandrevala) [inline src/ProtocolHandler/CzoTelNetHandler.cs]: Think of a better name than `Handle`

File: src/ProtocolHandler/CzoTelNetHandler.cs [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+        }
+
+        private string retrieveData(ICzoSocket chorizoSocket)
+        {
+            var bufferText = "";
+            var receivedData = new byte[0];
+            
+            while (bufferText.IndexOf("\n") == -1)
+            {
+                var (data, dataLength) = chorizoSocket.Receive(1);
```
Comment (cmvandrevala) [inline src/ProtocolHandler/CzoTelNetHandler.cs]: I might be wrong, but there might be a cleaner way of doing this

File: src/ProtocolHandler/CzoTelNetHandler.cs [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+                var originalLength = receivedData.Length;
+                Array.Resize(ref receivedData, originalLength + dataLength);
+                Array.Copy(data, 0, receivedData, originalLength, dataLength);
+                bufferText = Encoding.ASCII.GetString(receivedData, 0, receivedData.Length);
+            }
+            Console.WriteLine($"Got Data: {bufferText}");
```
Comment (cmvandrevala) [inline src/ProtocolHandler/CzoTelNetHandler.cs]: I might hide Console.WriteLine in an IO class

File: src/Sockets/CzoSocket/ICzoSocket.cs [hunk 13]
```diff
@@ -0,0 +1,12 @@
+using System;
+using System.Net.Sockets;
+
+namespace Chorizo.Sockets.CzoSocket
+{
+    public interface ICzoSocket {
+        Tuple<byte[], int> Receive(int byteBufferSize);
+        int Send(byte[] buffer);
+        void Shutdown(SocketShutdown how);
+        void Close();
```
Comment (cmvandrevala) [inline src/Sockets/CzoSocket/ICzoSocket.cs]: It is a bit weird to see both Shutdown and Close

Comment (cmvandrevala) [reply src/Sockets/CzoSocket/ICzoSocket.cs]: But that is how it is in C#

File: tests/ChorizoServerTest.cs [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+                .Returns(true)
+                .Returns(false);
+            _mockHTTPHandler = new Mock<ICzoProtocolHandler>();
+            _mockHTTPHandler.Setup(http => http.WillHandle("HTTP")).Returns(true);
+        }
+
+        [Fact]
+        public void Start_ShouldStartListeningOnDefaultPortAndHostNameUsingSocketMachine()
+        {
+            // Arrange
```
Comment (cmvandrevala) [inline tests/ChorizoServerTest.cs]: You don't really need these comments

File: src/ChorizoServer.cs [hunk 15]
```diff
@@ -11,26 +11,18 @@ public class Chorizo
...
-            SocketMachine.Configure(Config.Port, Config.HostName);
-        }
-
         public Chorizo(
-            int port,
-            string protocol,
-            IServerStatus serverStatus,
-            ISocketMachine socketMachine,
-            IChorizoProtocolConnectionHandler protocolConnectionHandler
+            int port = 8000,
```
Comment (ssunday) [inline src/ChorizoServer.cs]: Very nice

File: src/Constants.cs [hunk 16]
```diff
@@ -0,0 +1,9 @@
+namespace Chorizo
+{
+    public struct Constants
+    {
+        public const string HostName = "localhost";
+        public const string ServerMode = "dev";
```
Comment (ssunday) [inline src/Constants.cs]: 👏

**Pull request merged**